File: Strings/slides/BS-2/subs.hs
import Data.Char
import Data.Set

p = 31
q = 10^9 + 7

f :: Char -> Int
f c = (ord c) - (ord 'a') + 1

h :: String -> Int
h s = sum (zipWith (*) fs ps) `mod` q where
    fs = Prelude.map f s
    ps = Prelude.map (\x -> p ^ x) $ take (length s) [0..]

prefixes :: String -> [Int]
prefixes s = [h $ take k s | k <- [1..n]] where n = length s

fastExpMod :: Int -> Int -> Int
fastExpMod _ 0 = 1
fastExpMod a n = (b * fastExpMod (a^2 `mod` q) (n `div` 2)) `mod` q where
    b = if n `mod` 2 == 1 then a else 1

inverses :: Int -> [Int]
inverses n = [fastExpMod (fastExpMod p i) (q - 2) | i <- [0..(n-1)]]

hsubs i j ps is
    | i == 0 = ps !! j
    | otherwise = (ps !! j - ps !! (i - 1)) * is !! i `mod` q 

unique_substrings s = length us where
    n = length s
    xs = [(i, j) | i <- [0..(n-1)], j <- [i..(n-1)]] 
    ps = prefixes s
    is = inverses n
    hs = [hsubs i j ps is | (i, j) <- xs]
    us = fromList hs        -- us :: Data.Set

main = do
    print $ unique_substrings "tep"
    print $ unique_substrings "banana"
    print $ unique_substrings "aaaaa"

File: Strings/slides/hashes/codes/subs.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;
constexpr ll p = 31, q = 1'000'000'007;

int f(char c)
{
    return c - 'a' + 1;
}

int h(const string& s)
{
    ll ans = 0;

    for (auto c : s)
    {
        ans = (ans * p) % q;
        ans = (ans + f(c)) % q;
    }

    return ans;
}

vector<ll> prefixes(const string& s)
{
    int N = s.size();
    vector<ll> ps(N, 0);

    for (int i = 0; i < N; ++i)
        ps[i] = h(s.substr(0, i + 1));

    return ps;
}

ll fast_exp_mod(ll a, ll n)
{
    ll res = 1, base = a;

    while (n) {
        if (n & 1)
            res = (res * base) % q;

        base = (base * base) % q;
        n >>= 1;
    }

    return res;
} 

vector<ll> inverses(ll N)
{
    vector<ll> is(N);
    ll base = 1;

    for (int i = 0; i < N; ++i)
    {
        is[i] = fast_exp_mod(base, q - 2);
        base = (base * p) % q;
    }

    return is;
}

int h(int i, int j, const vector<ll>& ps, const vector<ll>& is)
{
    auto diff = i ? ps[j] - ps[i - 1] : ps[j];
    diff = (diff * is[i]) % q;

    return (diff + q) % q;
}

int unique_substrings(const string& s)
{
    set<ll> hs;
    int N = s.size();

    auto ps = prefixes(s);
    auto is = inverses(s.size());

    for (int i = 0; i < N; ++i)
    {
        for (int j = i; j < N; ++j)
        {
            auto hij = h(i, j, ps, is);
            hs.insert(hij);
        }
    }

    return hs.size();
}

int main()
{
    cout << unique_substrings("tep") << '\n';
    cout << unique_substrings("banana") << '\n';
    cout << unique_substrings("aaaaa") << '\n';

    return 0;
}
